architecture-docs/adr-001.md

```diff
@@ -0,0 +1,85 @@
+# ADR 001: Storage Consolidation
+
+## Changelog
+* 10-12-2019: Initial Draft
+
+## Context
+Currently (not counting Tendermint's internal storage or wallets), two processes maintain their internal storage:
+
+* chain-abci: stores the node state, Merkle trie of staking states, transaction metadata (whether spent or not),
+validator tracking, etc.
+* tx-validation enclave (TVE): sealed transaction data (of valid obfuscated transactions that have outputs)
+
+The reason for having two processes is that SGX SDK compilation is different and needs Intel SGX SDK tooling,
+so for the development convenience, the transaction validation code that needs to execute in an enclave
+is isolated. The inter-process communication is over a simple REQ-REP 0MQ socket.
```

Comment:
The only reason for separation of chain-abci and enclave is the difficulty of building enclave?

Reply:
yes

Reply:
I may clarify that in ADR: it's not only building, but also execution in development, as one can build and run chain-abci on any platform (e.g. macOS), and run the enclave stuff inside a docker container or on a remote Linux host.

That being said, it would be possible to remove these "enclave wrapper app" processes (like https://github.com/crypto-com/chain/tree/master/chain-tx-enclave/tx-validation/app) processes and move all that to chain-abci. It'd involve roughly this:

1. the enclave part will still need to be compiled to a static library as before + use the Intel SDK tooling (`sgx_sign`, `sgx_edger8r`, `compiler-rt`) to create a signed library, C-shim headers and files and compile them.

2.  chain-abci will need to expose these and its build process will need to be modify to link against these C-shims -- something like https://github.com/crypto-com/chain/blob/master/chain-tx-enclave/tx-validation/app/build.rs#L33 not sure if it can work with multiple enclaves

3. chain-abci will need to run in an environment with AESM and launch these enclaves like: https://github.com/crypto-com/chain/blob/master/chain-tx-enclave/enclave-u-common/src/enclave_u/mod.rs#L47 (perhaps instead of a hardcoded file, it'd need to accept paths to the signed static binaries of different enclaves, like txvalidation.signed.so, txquery.signed.so ...)

--
With that approach:

Positives:
* only one process

Negatives:
* inconveniences in development

Reply:
@yihuang @leejw51crypto @devashishdxt any thoughts on that?

Reply:
only one process seems like a big win, I'm not sure how much we can mitigate the development inconveniences, with current `mock-enc-dec` feature, we can already develop big part of project without enclave.

Reply:
@yihuang Do you want to merge codebases of run whole `chain-abci` in enclave?

Reply:
> @yihuang Do you want to only merge codebases of run whole `chain-abci` in enclave?

I needs more readings to comment on this, you can ignore my last comment I guess ;D

Reply:
It may be possible to have `build.rs` that with `mock-enc-dec`, it won't do anything, otherwise it'll call the linkers: https://github.com/crypto-com/chain/blob/master/chain-tx-enclave/tx-validation/app/build.rs#L33